### DISCUSSION TITLE
avrdude error: no usb support, please compile again with libusb installed

### DISCUSSION THREAD

**Discussion by lixtheyt** [avrdudes/avrdude#1685] [Q&A] [answered]
Hello! I am new to this but I wanted to program my m328ú (ATMega328p) using this command:
`avrdude -c ch341a -p238p -P COM4 -U flash:r:flash_dump.hex:i`
but i got this errors:
```
avrdude error: no usb support, please compile again with libusb installed
avrdude error: unable to open port COM4 for programmer ch341a

avrdude done.  Thank you.

```

Can somebody help me please?
I tried to build it but I got like thousands of error because of Windows SDK (i got installed) that it can't find packages and blah blah.

Thank you!

**Comment by mcuee** [comment]
Please read the FAQ on how to report an issue.
https://github.com/avrdudes/avrdude/wiki/FAQ#how-to-report-an-issue-to-avrdude-project

As per the FAQ, please also read the know limitation and see if you can figure out the issue.
https://github.com/avrdudes/avrdude/wiki/Known-limitations-of-avrdude

**Comment by lixtheyt** [comment]
OS : Windows 11
Programmer : CH341A
Part : ??? (i want to program m328p)
How host, programmer and part are connected/rigged up : Ill send
AVRDUDE binary version used (in bash avrdude -v 2>&1 | grep -i version) : 7.3-20240208
Full command line including all options : avrdude -c ch341a -p m328p -P COM4 -U flash:r:flash_backup.hex:i
Output of the command : 
```
avrdude warning: cannot open USB device: Permission denied
avrdude error: could not find USB device with vid=0x1a86 pid=0x5512
avrdude error: unable to open port COM4 for programmer ch341a

avrdude done.  Thank you.
```
Please try to use the latest released version : Tried
You may be requested to try out git main as well : I am idiot so i dont know how to
Sometimes it also helps us if you upload a file with the -vv or -vvvv output : idk what is that
Sometimes it also helps us if you post the hex file used (if you are okay to share the hex file) : what?
What you would have expected to happen : create file with that what is in m328p 💀

**Comment by lixtheyt** [comment]
![ch341a_miniprogrammer](https://github.com/avrdudes/avrdude/assets/129482397/8209a9ce-deee-44f3-9722-e0b57d0cd98e)
![image002](https://github.com/avrdudes/avrdude/assets/129482397/d875cd13-b140-450e-8cc8-9ea35b524f62)

**Reply by mcuee** [reply, marked as answer]
@lixtheyt 

Please also read the known limitation of the programmer here. CH341A is not supported by the MSVC build.
https://github.com/avrdudes/avrdude/wiki/Known-limitations-of-avrdude
* https://github.com/avrdudes/avrdude/discussions/1440
* https://github.com/avrdudes/avrdude/discussions/1441

CH341A is not well supported by avrdude. I will recommend you to go with another programer like usbasp.

If you want to continue, please try the mingw64 build here. You have to use Zadig to change the vendor driver to WinUSB as well.
https://github.com/avrdudes/avrdude/wiki/Getting-Nightly-Builds-for-AVRDUDE
https://nightly.link/avrdudes/avrdude/workflows/build/main/avrdude-mingw-x86_64.zip

**Reply by lixtheyt** [reply]
Okay I tried and its working but when I want to write something it writes this:

```
avrdude: device signature = 0x000000 (retrying)
avrdude: device signature = 0x000000 (retrying)
avrdude: device signature = 0x000000
avrdude error: Yikes!  Invalid device signature.
avrdude warning: expected signature for ATmega328P is 1E 95 0F

```

and I tried to rewrite signature and this happend:

```
avrdude error: initialization failed, rc=-2
        the programmer ISP clock is too fast for the target
        - use -B to set lower the bit clock frequency, e.g. -B 125kHz
avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x000000 (retrying)
avrdude: device signature = 0x000000 (retrying)
avrdude: device signature = 0x000000
avrdude error: Yikes!  Invalid device signature.
avrdude warning: expected signature for ATmega328P is 1E 95 0F

avrdude: processing -U signature:w:0x1e,0x95,0x0f:m
avrdude: reading input file 0x1e,0x95,0x0f for signature
         with 3 bytes in 1 section within [0, 2]
avrdude: writing 3 bytes signature ...
avrdude error: cannot write to read-only memory signature of ATmega328P
 ***failed;
avrdude error: cannot write to read-only memory signature of ATmega328P
 ***failed;
avrdude error: cannot write to read-only memory signature of ATmega328P
 ***failed;
avrdude: 3 bytes of signature written
avrdude: verifying signature memory against 0x1e,0x95,0x0f
avrdude warning: verification mismatch
        device 0xff != input 0x1e at addr 0x0000 (error)
avrdude error: verification mismatch

avrdude done.  Thank you.
```

btw I tried -B 1kHz and still it wrote that is too fast so thats why I used -F :)
What can I do ? THX

**Reply by MCUdude** [reply]
The CH341A is too fast for a stock AVR, even if you reduce the clock speed using -B.
You'll have to use a different programmer to change the default fuses in order to run the AVR at minimum 6.8 MHz.

https://github.com/avrdudes/avrdude/blob/0179026e09d19833e7dea06ed8db19f971138e4b/src/avrdude.conf.in#L2138-L2169

**Comment by lixtheyt** [comment]
I conncted SCK with CLK and RESET with CS

**Comment by stefanrueger** [comment]
Good progress: you got over the "compile again with libusb installed" hurdle. Now there is a permission problem. 
```
avrdude warning: cannot open USB device: Permission denied
```
Does another program use the ch341a programmer? That could interfere with avrdude, so close that program if so. Maybe it is as simple as giving COM4 read/write permission (this is an OS thing: I don't know how you would do that in Windows).

You get more info if you run your command with `-v`, `-vv`, `-vvv`, but I doubt this will show significantly different info

**Comment by lixtheyt** [comment]
Thank you
